File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root) return NULL;
        if(root==p || root==q) return root;
        TreeNode* lst = lowestCommonAncestor(root->left,p,q);
        TreeNode* rst = lowestCommonAncestor(root->right,p,q); 
        if(lst && rst) return root; 
        if(lst && !rst) return lst;
        return rst;
    }
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None: 
            return None 
        if root is p or root is q:
            return root 

        lst = self.lowestCommonAncestor(root.left,p,q)
        rst = self.lowestCommonAncestor(root.right,p,q)
        
        if lst and (rst is None): 
            return lst 
        if rst and (lst is None): 
            return rst 
        if lst and rst:
            return root
        return None